modules/bluefish/consumer/bluefish_consumer.cpp

```diff
@@ -237,15 +463,77 @@ struct bluefish_consumer : boost::noncopyable
 		});
 	}
 
-	void display_frame(core::const_frame frame)
+	void dma_present_thread_actual()
 	{
-		// Sync
+		bvc_wrapper wait_b;
+		wait_b.attach(device_index_);
+		EBlueVideoChannel out_vid_channel = get_bluesdk_videochannel_from_streamid(device_output_channel_);
+		wait_b.set_card_property32(DEFAULT_VIDEO_OUTPUT_CHANNEL, out_vid_channel);
+		int frames_to_buffer = BLUEFISH_HW_BUFFER_DEPTH;
+		unsigned long buffer_id = 0;
+		unsigned long underrun = 0;
+
+		while (!end_dma_thread_)
+		{
+			if (!live_frames_.empty() && BLUE_OK(blue_->video_playback_allocate(buffer_id, underrun)))
```

Comment:
Actually I am not sure if `tbb::concurrent_bounded_queue::empty` is guaranteed to give the expected result. It is based on `size()` and that doesn't seem that convincing in the api docs:
> Number of pushes minus number of pops. The result is negative if there are pop operations waiting for corresponding pushes. The result can exceed capacity() if the queue is full and there are push operations waiting for corresponding pops."

The correct way is to use the `try_pop()` method, which pops if an item is available and returns `true` otherwise `false`. You can see example uses of that method in other places in the CasparCG Server code.